### PR TITLE
Added AutoEnsembleForecaster and refactor/doc of StackingForecaster

### PR DESCRIPTION
**Pull request by aiwalter**
#### Reference Issues/PRs
Issue #1214 

#### What does this implement/fix? Explain your changes.
Automatically finding best weights for ensemble based on a meta-model (regressor). Weights are taken from
either `regressor.coef_` for all `sklearn` linear models or from `regressor.feature_importances_` for all decision tree-based models.

I already proved in an example that `AutoEnsembleForecaster` is outperforming each single given forecaster for `load_airline` data, so we know that this ensemble approach can outperform. Code is here to see the error differences:
```python
y = load_airline()
y_train, y_test = temporal_train_test_split(y)
fh = ForecastingHorizon(y_test.index, is_relative=False)

# create ensemble
forecasters = [
    ("naive", NaiveForecaster(strategy="drift")),
    ("ets", AutoETS(auto=True, n_jobs=-1, sp=12)),
    ("arima", AutoARIMA(suppress_warnings=True)),
    ("bats", BATS(sp=12, use_trend=True)),
    ("tbats", TBATS(sp=12, use_trend=True))
]
auto = AutoEnsembleForecaster(forecasters=forecasters)
auto.fit(y_train, fh=fh)

# get prediction and weights
y_weighted = auto.predict(fh)
auto.weights_
> [0.0046995889455789235,
> 0.4021907840860222,
> 0.01191944521686898,
> 0.12701615072179243,
> 0.4541740310297374]

# create benchmark forecasters, same as given to ensemble
forecaster_1 = NaiveForecaster(strategy="drift")
forecaster_2 = AutoETS(auto=True, n_jobs=-1, sp=12)
forecaster_3 = AutoARIMA(suppress_warnings=True)
forecaster_4 = BATS(sp=12, use_trend=True)
forecaster_5 = TBATS(sp=12, use_trend=True)

y_pred_1 = forecaster_1.fit(y_train).predict(fh)
y_pred_2 = forecaster_2.fit(y_train).predict(fh)
y_pred_3 = forecaster_3.fit(y_train).predict(fh)
y_pred_4 = forecaster_4.fit(y_train).predict(fh)
y_pred_5 = forecaster_5.fit(y_train).predict(fh)

# get benchmark results
print("naive \t", mean_squared_error(y_test, y_pred_1))
print("ets \t", mean_squared_error(y_test, y_pred_2))
print("arima \t", mean_squared_error(y_test, y_pred_3))
print("bats \t", mean_squared_error(y_test, y_pred_4))
print("tbats \t", mean_squared_error(y_test, y_pred_5))
print("ens \t", mean_squared_error(y_test, y_weighted))

> naive 	 7695.698285148631
> ets 	 1153.7492147391474
> arima 	 3505.3729027077143
> bats 	 887.3656338724588
> tbats 	 801.3102613074158
> ens 	 584.1444166907737
```
Plot of the forecasts:
output

This PR also contains a refactor of `StackingForecaster` and docs.

#### What should a reviewer concentrate their feedback on?
Any ideas to improve UX and performance of this class?

##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/alan-turing-institute/sktime/blob/main/.all-contributorsrc).
- [x] Optionally, I've updated sktime's [CODEOWNERS](https://github.com/alan-turing-institute/sktime/blob/main/CODEOWNERS) to receive notifications about future changes to these files.
- [x] I've added unit tests and made sure they pass locally.

##### For new estimators
- [x] I've added the estimator to the online documentation.
- [ ] I've updated the existing example notebooks or provided a new one to showcase how my estimator works.

### CODE REVIEW

**Comment by aiwalter** [comment]
@RNKuhns @GuzalBulatova could you pls review?

File: sktime/forecasting/base/_meta.py
Comment (GuzalBulatova) [inline sktime/forecasting/base/_meta.py:98]: Just a reminder, once [#1190](https://github.com/alan-turing-institute/sktime/pull/1190/commits) is merged, `_update` will be in `_HeterogenousEnsembleForecaster`.

File: sktime/forecasting/compose/_ensemble.py
Comment (GuzalBulatova) [inline sktime/forecasting/compose/_ensemble.py:36]: typo - optional

**Review by GuzalBulatova — COMMENTED** [review]
Looks all fine to me! Good idea to rename `final_regressor` in `StackingForecaster`. 👍

File: sktime/forecasting/base/_meta.py
Comment (aiwalter) [reply sktime/forecasting/base/_meta.py:98]: yes this is `_HeterogenousEnsembleForecaster`, I moved it there also. It will be fine with merging

**Comment by aiwalter** [comment]
I will fix `pydocstyle` after #1162 is merged.

File: sktime/forecasting/compose/_ensemble.py
Comment (GuzalBulatova) [inline sktime/forecasting/compose/_ensemble.py:5]: the author string should have your GitHub username, so it'll be "aiwalter"

File: sktime/forecasting/compose/_ensemble.py
Comment (GuzalBulatova) [inline sktime/forecasting/compose/_ensemble.py:54]: Examples
--------

File: sktime/forecasting/compose/_ensemble.py
Comment (aiwalter) [reply sktime/forecasting/compose/_ensemble.py:5]: actually we should maybe replace them all everywhere in a separate PR

**Comment by aiwalter** [comment]
@RNKuhns @GuzalBulatova @mloning @fkiraly pls kindly review this.

**Review by GuzalBulatova — APPROVED** [review]
Looks all very good to me, great job, @aiwalter! The code is very well written and easy to read.
What do you think about adding some simple tests for AutoEnsemble, maybe in another PR?

**Comment by aiwalter** [comment]
thx @GuzalBulatova :)

> What do you think about adding some simple tests for AutoEnsemble, maybe in another PR?

I thought about it, but was not sure what exactly we could test besides the tests that already exist for all forecasters. Any ideas? Otherwise pls approve :)

**Comment by aiwalter** [comment]
@fkiraly could you in return have a look on this PR? :) 
@GuzalBulatova pls approve again, sry.

File: sktime/forecasting/base/_meta.py
Comment (fkiraly) [inline sktime/forecasting/base/_meta.py:33]: question: why does this need to be in the base class?
To me, it seems this is a concern that is very specific to stacking.
I wouldn't put it here, all the other descendants get it but don't need it, it seems to me?

File: sktime/forecasting/compose/_ensemble.py
Comment (fkiraly) [inline sktime/forecasting/compose/_ensemble.py:120]: hm, do we want to upgrade `temporal_train_test_split` so it returns `None` for the `X`-type returns if `X` is `None`? Would make this code (and probably other code) easier.

Comment (fkiraly) [reply sktime/forecasting/compose/_ensemble.py:120]: ah, that wouldn't work, since it would break the expected two arguments

File: sktime/forecasting/compose/_ensemble.py
Comment (fkiraly) [inline sktime/forecasting/compose/_ensemble.py:163]: this is now handled by the base class

File: sktime/forecasting/base/_meta.py
Comment (fkiraly) [reply sktime/forecasting/base/_meta.py:33]: ok, got it now: there are two descendants that need it, stacking and the new estimator.

File: sktime/transformations/series/compose.py
Comment (fkiraly) [inline sktime/transformations/series/compose.py:83]: what is this? Should this not be in a different PR?

**Comment by fkiraly** [comment]
looks good, happy to approve once tests pass.

There's a change to `OptionalPassthrough`, seems like an accident as it should be in a different PR?

File: sktime/transformations/series/compose.py
Comment (aiwalter) [reply sktime/transformations/series/compose.py:83]: I added it here to avoid spamming the repo with tiny PRs as CI/CD seems to be slow.

**Comment by aiwalter** [comment]
Seems to be ok now.

**Comment by aiwalter** [comment]
pls approve and merge @fkiraly

File: sktime/transformations/series/compose.py
Comment (fkiraly) [reply sktime/transformations/series/compose.py:83]: ok, fine

**Review by fkiraly — APPROVED** [review]
looks good to me, straightforward addition

moving `_check_regressor` to the base class also makes sense to me

**Comment by fkiraly** [comment]
this doesn't seem to add the forecaster to the docs, I assume that will be another PR?

**Comment by aiwalter** [comment]
oh, I will add this in a separate PR. Thx!

File: docs/source/api_reference/forecasting.rst
Comment (aiwalter) [inline docs/source/api_reference/forecasting.rst:141]: its there :) or what is missing @fkiraly ?

Comment (fkiraly) [reply docs/source/api_reference/forecasting.rst:141]: ah, it's actually there - forget what I said

**Pull request merged**
